### ISSUE TITLE
List extending panic

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#7190]
Migrated from [rt.perl.org#27811](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=27811) (status was 'resolved')

Searchable as RT27811$

**Comment by p5pRT** [comment]
### From andy-baby@petdance.com

#### Created by andy-baby@petdance.com



While adding tests on \#27791\, I came across this&#8203;:

$ perlblead \-le'@&#8203;x=\(1\);@&#8203;y=\(@&#8203;x\)x\-1'
panic&#8203;: list extend at \-e line 1

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.8.1:

Configured by andy at Thu Sep 25 22:12:05 CDT 2003.

Summary of my perl5 (revision 5.0 version 8 subversion 1) configuration:
  Platform:
    osname=linux, osvers=2.4.20-8, archname=i686-linux
    uname='linux baby 2.4.20-8 #1 thu mar 13 17:54:28 est 2003 i686 i686 i386 gnulinux '
    config_args='-des -Dprefix=/usr/local -Dcf_email=andy-baby@petdance.com -Dperladmin=andy-baby@petdance.com'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I/usr/include/gdbm',
    optimize='-O3',
    cppflags='-fno-strict-aliasing -I/usr/local/include -I/usr/include/gdbm'
    ccversion='', gccversion='3.2.2 20030222 (Red Hat Linux 3.2.2-5)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.3.2.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.3.2'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.1:
    /usr/local/lib/perl5/5.8.1/i686-linux
    /usr/local/lib/perl5/5.8.1
    /usr/local/lib/perl5/site_perl/5.8.1/i686-linux
    /usr/local/lib/perl5/site_perl/5.8.1
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.8.1:
    HOME=/home/andy
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/local/bin:/bin:/usr/bin:/usr/X11R6/bin:/home/andy/bin:/home/andy/play
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @petdance

\(Note this is faking the perlbug report\)

Repeating a list with "x \-1" \(or any negative number\) causes a panic\,
but only in blead\.

andy@&#8203;baby\[~\]$ cat testmany
\#\!/bin/bash

for VER in 5\.4\.0 5\.5\.3 5\.6\.0 5\.6\.1 5\.6\.2 5\.8\.0 5\.8\.1 5\.8\.2 5\.8\.3 5\.9\.1 blead; do
&nbsp;    PERL="perl$VER"
&nbsp;    echo $PERL
&nbsp;    $PERL \-le'@&#8203;x=1;@&#8203;y=\(@&#8203;x\)x\-1'
done


andy@&#8203;baby\[~\]$ \./testmany
perl5\.4\.0
perl5\.5\.3
perl5\.6\.0
perl5\.6\.1
perl5\.6\.2
perl5\.8\.0
perl5\.8\.1
perl5\.8\.2
perl5\.8\.3
perl5\.9\.1
perlblead
panic&#8203;: list extend at \-e line 1\.


Does it make sense to x\-1?  Personally\, I think a negative number ought
to act like zero\, but whatever it does\, we should document it\.

xoa

\-\- 
Andy Lester => andy@&#8203;petdance\.com => www\.petdance\.com => AIM&#8203;:petdance

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @andk

>>>>> On Sat\, 20 Mar 2004 21&#8203;:05&#8203;:36 \-0600\, Andy Lester \<andy@&#8203;petdance\.com> said&#8203;:

\-\-\-\-Program\-\-\-\-
@&#8203;x=1;@&#8203;y=\(@&#8203;x\)x\-1

\-\-\-\-Output of \.\.\./pQb9Nqu/perl\-5\.8\.0@&#8203;22516/bin/perl\-\-\-\-

\-\-\-\-EOF \($?='0'\)\-\-\-\-
\-\-\-\-Output of \.\.\./pxEitd4/perl\-5\.8\.0@&#8203;22517/bin/perl\-\-\-\-
panic&#8203;: list extend at tests/andy\_x\-1\.pl line 1\.

\-\-\-\-EOF \($?='65280'\)\-\-\-\-

Change 22517 by nicholas@&#8203;faith on 2004/03/17 17&#8203;:58&#8203;:44

&nbsp;        Subject&#8203;: malloc wrappage
&nbsp;        From&#8203;: Jarkko Hietaniemi \<jhi@&#8203;iki\.fi> 
&nbsp;        Message\-Id&#8203;: \<37BF70DE\-5C0E\-11D8\-B5A1\-00039362CB92@&#8203;iki\.fi>
&nbsp;        Date&#8203;: Tue\, 10 Feb 2004 23&#8203;:15&#8203;:14 \+0200
&nbsp;        
&nbsp;        plus change croak to Perl\_croak\_nocontext to make ithread safe
&nbsp;        plus make it conditional on PERL\_MALLOC\_WRAP \(default for blead is on\)



\-\- 
andreas

**Comment by p5pRT** [comment]
### From @rgs

Andy Lester wrote&#8203;:
>     $PERL \-le'@&#8203;x=1;@&#8203;y=\(@&#8203;x\)x\-1'
\<
> Does it make sense to x\-1?  Personally\, I think a negative number ought
> to act like zero\, but whatever it does\, we should document it\.

This should at the very least produce a run time warning\.

**Comment by p5pRT** [comment]
### From @rgs

Rafael Garcia\-Suarez wrote&#8203;:
>
> Andy Lester wrote&#8203;:
> >     $PERL \-le'@&#8203;x=1;@&#8203;y=\(@&#8203;x\)x\-1'
> \<
> > Does it make sense to x\-1?  Personally\, I think a negative number ought
> > to act like zero\, but whatever it does\, we should document it\.
> 
> This should at the very least produce a run time warning\.

\.\.\.which I added as \#22543\. \(and I fixed the panic too\)

**Comment by p5pRT** [comment]
@rgs - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @petdance

> > Does it make sense to x\-1?  Personally\, I think a negative number ought
> > to act like zero\, but whatever it does\, we should document it\.
> 
> This should at the very least produce a run time warning\.

I don't disagree\.  However\, as it stands now\, and as it has been back to
5\.004\, a negative number is effectively a zero\.  We should apply my
patch and Richard's patch to at least get it documented and non\-panicing
while p5p argues at length about it\.

xoa

\-\- 
Andy Lester => andy@&#8203;petdance\.com => www\.petdance\.com => AIM&#8203;:petdance

**Comment by p5pRT** [comment]
### From @petdance

> I don't disagree\.  However\, as it stands now\, and as it has been back to
> 5\.004\, a negative number is effectively a zero\.  We should apply my
> patch and Richard's patch to at least get it documented and non\-panicing
> while p5p argues at length about it\.

And we still need the original test patch applied\.

<details><summary>Inline Patch</summary>

```diff
--- bleadperl/t/op/repeat.t	2001-11-18 12:11:25.000000000 -0600
+++ patch/t/op/repeat.t	2004-03-20 19:07:45.000000000 -0600
@@ -6,13 +6,15 @@
 }
 
 require './test.pl';
-plan(tests => 25);
+plan(tests => 33);
 
 # compile time
 
 is('-' x 5, '-----',    'compile time x');
 is('-' x 1, '-',        '  x 1');
 is('-' x 0, '',         '  x 0');
+is('-' x -1, '',        '  x -1');
+is('-' x undef, '',     '  x undef');
 
 is('ab' x 3, 'ababab',  '  more than one char');
 
@@ -22,9 +24,15 @@
 is($a x 5, '-----',     'run time x');
 is($a x 1, '-',         '  x 1');
 is($a x 0, '',          '  x 0');
+is($a x -3, '',         '  x -3');
+is($a x undef, '',      '  x undef');
 
 $a = 'ab';
 is($a x 3, 'ababab',    '  more than one char');
+$a = 'ab';
+is($a x 0, '',          '  more than one char');
+$a = 'ab';
+is($a x -12, '',        '  more than one char');
 
 $a = 'xyz';
 $a x= 2;
@@ -45,6 +53,9 @@
 is(join(':', (9,9) x 4),    '9:9:9:9:9:9:9:9',      '(X,X) x Y');
 is(join('', (split(//,"123")) x 2), '123123',       'split and x');
 
+is(join('', @x x -12),      '',                     '@x x -12');
+is(join('', (@x) x -14),    '',                     '(@x) x -14');
+
 
 # This test is actually testing for Digital C compiler optimizer bug,
 # present in Dec C versions 5.* and 6.0 (used in Digital UNIX and VMS),

-- 
```
</details>

Andy Lester => andy@&#8203;petdance\.com => www\.petdance\.com => AIM&#8203;:petdance

**Comment by p5pRT** [comment]
### From @mjdominus

Rafael Garcia\-Suarez \<rgarciasuarez@&#8203;free\.fr>&#8203;:
> Rafael Garcia\-Suarez wrote&#8203;:
> >
> > Andy Lester wrote&#8203;:
> > >     $PERL \-le'@&#8203;x=1;@&#8203;y=\(@&#8203;x\)x\-1'
> > \<
> > > Does it make sense to x\-1?  Personally\, I think a negative number ought
> > > to act like zero\, but whatever it does\, we should document it\.
> > 
> > This should at the very least produce a run time warning\.
> 
> \.\.\.which I added as \#22543\. \(and I fixed the panic too\)

I wonder if it's really a good idea to put in a warning for this\.  It
seems to me like behavior someone might reasonably be depending on\.
It's analogous to the way that

&nbsp;        $x \.\. $y

yields the empty list when $x > $y\.  There's no warning for \.\. when 
$x > $y\, and 'x \-1' seems similar to me\.  Is it possible that people
are overreacting because of the panic?

**Comment by p5pRT** [comment]
### From perl5-porters@ton.iguana.be

In article \<20040321140746\.7003\_3a3\.rgarciasuarez@&#8203;\_ree\.\_r>\,
&nbsp;    Rafael Garcia\-Suarez \<rgarciasuarez@&#8203;free\.fr> writes&#8203;:
> Andy Lester wrote&#8203;:
>>     $PERL \-le'@&#8203;x=1;@&#8203;y=\(@&#8203;x\)x\-1'
> \<
>> Does it make sense to x\-1?  Personally\, I think a negative number ought
>> to act like zero\, but whatever it does\, we should document it\.
> 
> This should at the very least produce a run time warning\.
I at least used it in real code\. It's very convenient when e\.g\.
"almost" lining up\. Think&#8203;:
&nbsp;  $str \. " " x \($width \- length $str\) \. " Something that comes behind"
\(but then for more complex cases\, for this you could also use sprintf\)\.

**Comment by p5pRT** [comment]
### From @rgs

Andy Lester wrote&#8203;:
>
> > I don't disagree\.  However\, as it stands now\, and as it has been back to
> > 5\.004\, a negative number is effectively a zero\.  We should apply my
> > patch and Richard's patch to at least get it documented and non\-panicing
> > while p5p argues at length about it\.
> 
> And we still need the original test patch applied\.
> 
> \-\-\- bleadperl/t/op/repeat\.t    2001\-11\-18 12&#8203;:11&#8203;:25\.000000000 \-0600
> \+\+\+ patch/t/op/repeat\.t    2004\-03\-20 19&#8203;:07&#8203;:45\.000000000 \-0600

Thanks\, applied as \#22549\.

\(I have probably missed Richard's patch\, I don't see it anywhere\)

**Comment by p5pRT** [comment]
### From @rgs

Ton Hospel wrote&#8203;:
>
> In article \<20040321140746\.7003\_3a3\.rgarciasuarez@&#8203;\_ree\.\_r>\,
>     Rafael Garcia\-Suarez \<rgarciasuarez@&#8203;free\.fr> writes&#8203;:
> > Andy Lester wrote&#8203;:
> >>     $PERL \-le'@&#8203;x=1;@&#8203;y=\(@&#8203;x\)x\-1'
> > \<
> >> Does it make sense to x\-1?  Personally\, I think a negative number ought
> >> to act like zero\, but whatever it does\, we should document it\.
> > 
> > This should at the very least produce a run time warning\.
> I at least used it in real code\. It's very convenient when e\.g\.
> "almost" lining up\. Think&#8203;:
>   $str \. " " x \($width \- length $str\) \. " Something that comes behind"
> \(but then for more complex cases\, for this you could also use sprintf\)\.

Right\. Spurious warning backed out \(as apparently "x $negnumber" is not
really an error difficult to catch compared to a desired behaviour\.\)

**Comment by p5pRT** [comment]
### From @floatingatoll

On 2004\-03\-20 19&#8203;:05&#8203;:36 \-0800\, andy@&#8203;petdance\.com \(Andy Lester\) said&#8203;:

> \(Note this is faking the perlbug report\)
>
> Repeating a list with "x \-1" \(or any negative number\) causes a panic\, 
> but only in blead\.
\[snip\]
> panic&#8203;: list extend at \-e line 1\.
>
> Does it make sense to x\-1?  Personally\, I think a negative number 
> ought to act like zero\, but whatever it does\, we should document it\.

The code path in pp\_repeat is three\-pronged&#8203;: if count > 1\, then repeat 
the items count times; if count \< 1\, then remove the items; otherwise\, 
if count == 0\, leave the items alone\.  The "list extend" panic is 
triggered when the final count of items is calculated \-\- but it's being 
calculated \*before\* the count branch\.  Since the list panic is 
triggered by calculations performed with negative values of count\, 
moving the calculations inside the count > 1 block resolves the panic\.  
The results of the calculations are only necessary if the count > 1 
branch is taken\, so this bugfix is also technically an optimization :\)    

Here's a patch for the code\, to fix the panic; this will resolve the 
problem you described in \#27791\, as well; porters\, Andy's patch \[1\] 
provides tests and documentation for the now\-fixed behavior of 
pp\_repeat with negative counts\.

\[1\] http&#8203;://www\.nntp\.perl\.org/group/perl\.perl5\.porters/89887

&nbsp;  \- R\.

<details><summary>Inline Patch</summary>

```diff
--- perl-current/pp.c	Thu Mar 18 11:23:19 2004
+++ ./pp.c	Sat Mar 20 19:50:34 2004
@@ -1389,15 +1389,17 @@
      if (GIMME == G_ARRAY && PL_op->op_private & OPpREPEAT_DOLIST) {
  	dMARK;
  	I32 items = SP - MARK;
-	I32 max;
-	static const char list_extend[] = "panic: list extend";

-	max = items * count;
-	MEM_WRAP_CHECK_1(max, SV*, list_extend);
-	if (items > 0 && max > 0 && (max < items || max < count))
-	   Perl_croak(aTHX_ list_extend);
-	MEXTEND(MARK, max);
  	if (count > 1) {
+	    I32 max;
+	    static const char list_extend[] = "panic: list extend";
+
+	    max = items * count;
+	    MEM_WRAP_CHECK_1(max, SV*, list_extend);
+	    if (items > 0 && max > 0 && (max < items || max < count))
+	         Perl_croak(aTHX_ list_extend);
+	    MEXTEND(MARK, max);
+
  	    while (SP > MARK) {
  #if 0
  	      /* This code was intended to fix 20010809.028:
```
</details>

**Comment by p5pRT** [comment]
### From @floatingatoll

<details><summary>negative-repeat-panic.patch</summary>

```diff
--- perl-current/pp.c	Thu Mar 18 11:23:19 2004
+++ ./pp.c	Sat Mar 20 19:50:34 2004
@@ -1389,15 +1389,17 @@
     if (GIMME == G_ARRAY && PL_op->op_private & OPpREPEAT_DOLIST) {
 	dMARK;
 	I32 items = SP - MARK;
-	I32 max;
-	static const char list_extend[] = "panic: list extend";
 
-	max = items * count;
-	MEM_WRAP_CHECK_1(max, SV*, list_extend);
-	if (items > 0 && max > 0 && (max < items || max < count))
-	   Perl_croak(aTHX_ list_extend);
-	MEXTEND(MARK, max);
 	if (count > 1) {
+	    I32 max;
+	    static const char list_extend[] = "panic: list extend";
+
+	    max = items * count;
+	    MEM_WRAP_CHECK_1(max, SV*, list_extend);
+	    if (items > 0 && max > 0 && (max < items || max < count))
+	         Perl_croak(aTHX_ list_extend);
+	    MEXTEND(MARK, max);
+
 	    while (SP > MARK) {
 #if 0
 	      /* This code was intended to fix 20010809.028:

```
</details>
